### PR TITLE
Rekovic hlt 80x

### PR DESCRIPTION
**Pull request by rekovic**
Stage2: From L1T emulation to HLTL1 interface (HLTL1TSeed)

### CODE REVIEW

**Comment by slava77** [comment]
data files in EventFilter/L1TRawToDigi
should go to
https://github.com/cms-data/EventFilter-L1TRawToDigi
and not to cmssw repository.

**Comment by fwyzard** [comment]
hi @slava77 , does it mean one needs to make a new external for the 4855 data files (42 MB) under L1Trigger/L1TMuon/data ?

What about the 15 or 16 files under L1Trigger/L1TCalorimeter/data ?

**Comment by fwyzard** [comment]
@rekovic , about the conflicts, 
  - it looks like part of the commit history (e.g. #1b5f295 , #593ae69 ) is duplicated between this branch and CMSSW_8_0_X ;
  - #58ce0b3 seems to have some issue in DataFormats/HLTReco/src/classes_def.xml .

**Comment by slava77** [comment]
On 2/1/16 2:00 PM, Andrea Bocci wrote:
> hi @slava77 <https://github.com/slava77> , does it mean one needs to
> make a new external for the 4855 data files (42 MB) under
> L1Trigger/L1TMuon/data ?
> 
> What about the 15 or 16 files under L1Trigger/L1TCalorimeter/data ?
> 

yes, large files that do not require compilation should be placed
in external packages.
Their best eventual destination is the database.

**Comment by mulhearn** [comment]
Folks, this is not a real PR.  This is for convenient of HLT folks to have all commits in one place for testing purposes.  The actual PRs are factored into smaller parts and don't include data files.

File: DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h [hunk 1]
```diff
@@ -62,7 +63,8 @@ namespace trigger
     
     /// collectionTags
     void addCollectionTag(const edm::InputTag& collectionTag){
-      collectionTags_.push_back(collectionTag.encode());
+      std::string tmp = collectionTag.encode();
+      collectionTags_.push_back(tmp);
     }
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h]: Why this change? It seems useless, so please revert.

File: DataFormats/HLTReco/interface/TriggerRefsCollections.h [hunk 2]
```diff
@@ -58,11 +64,17 @@ namespace trigger
...
+  typedef std::vector<l1extra::L1MuonParticleRef>           VRl1muon; //deprecate
+  typedef std::vector<l1extra::L1JetParticleRef>            VRl1jet; //deprecate
+  typedef std::vector<l1extra::L1EtMissParticleRef>         VRl1etmiss; //deprecate
+  typedef std::vector<l1extra::L1HFRingsRef>                VRl1hfrings; //deprecate
+
+  typedef l1t::MuonVectorRef                                VRl1tmuon;
+  typedef l1t::EGammaVectorRef                              VRl1tegamma;
+  typedef l1t::JetVectorRef                                 VRl1tjet;
+  typedef l1t::TauVectorRef                                 VRl1ttau;
+  typedef l1t::EtSumVectorRef                               VRl1tetsum;
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/interface/TriggerRefsCollections.h]: These new ones must have l1 in their name, as for all l1 types in here.

File: DataFormats/HLTReco/interface/TriggerRefsCollections.h [hunk 3]
```diff
@@ -162,6 +191,10 @@ namespace trigger
       std::swap(l1hfringsIds_,  other.l1hfringsIds_);
       std::swap(l1hfringsRefs_, other.l1hfringsRefs_);
 
+      std::swap(l1tmuonIds_,     other.l1tmuonIds_);
+      std::swap(l1tegammaIds_,   other.l1tegammaIds_);
+      std::swap(l1tjetIds_,      other.l1tjetIds_);
+
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/interface/TriggerRefsCollections.h]: You always add 5 new L1 types, but here there are only three.

File: DataFormats/HLTReco/src/classes_def.xml [hunk 4]
```diff
@@ -65,7 +79,14 @@
...
-  <class name="trigger::TriggerEventWithRefs" ClassVersion="12">
+  <class name="trigger::TriggerEventWithRefs" ClassVersion="19">
+   <version ClassVersion="19" checksum="1258968436"/>
+   <version ClassVersion="18" checksum="742506204"/>
+   <version ClassVersion="17" checksum="4163362996"/>
+   <version ClassVersion="16" checksum="1580092994"/>
+   <version ClassVersion="15" checksum="3347721344"/>
+   <version ClassVersion="14" checksum="1580092994"/>
+   <version ClassVersion="13" checksum="1828447080"/>
    <version ClassVersion="12" checksum="3347721344"/>
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/src/classes_def.xml]: Why do we need several generations of added class versions?
From your intermediate tests? We should just have the most recent ones as the intermediate ones are not used.

File: DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h [hunk 5]
```diff
@@ -24,6 +24,7 @@
 #include <string>
 #include <vector>
 #include <algorithm>
+#include <iostream>
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h]: Also this is not needed.

**Comment by Martin-Grunewald** [comment]
BTW, what is the status of the HLT trigger summary producers wrt the new L1T objects?

**Comment by rekovic** [comment]
@Martin-Grunewald  Thanks.
The HLT trigger summary producer has not been adjusted to the new changes yet in this PR.  It will be worked on now.

File: DataFormats/HLTReco/interface/TriggerRefsCollections.h [hunk 6]
```diff
@@ -754,7 +836,81 @@ namespace trigger
...
+      refs.resize(n);
+      size_type j(0);
+      for (size_type i=begin; i!=end; ++i) {
+	if (id==l1tetsumIds_[i]) {refs[j]=l1tetsumRefs_[i]; ++j;}
+      }
+      return;
+    }
     void getObjects(Vids& ids, VRpfjet& refs) const {
       getObjects(ids,refs,0,pfjetIds_.size());
     }
```
Comment (Martin-Grunewald) [inline DataFormats/HLTReco/interface/TriggerRefsCollections.h]: You also need to add the low-level accessors which are here:
  https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/DataFormats/HLTReco/interface/TriggerRefsCollections.h#L854

**Comment by davidlange6** [comment]
would be great if it were labeled that way:)

File: DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h (hunk 1 above)
Comment (mulhearn) [inline DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h]: will do (was for debugging print of tmp)

File: DataFormats/HLTReco/src/classes_def.xml (hunk 4 above)
Comment (mulhearn) [inline DataFormats/HLTReco/src/classes_def.xml]: they can go... the automatic tool that generates the checksum simply appends.

**Comment by fwyzard** [comment]
hi @Martin-Grunewald 
> BTW, what is the status of the HLT trigger summary producers wrt the new L1T objects?
is this something that you and @perrotta could work on ? looks like Mike and Vladmir are quite busy fixing the rest :-(

**Comment by rekovic** [comment]
@Martin-Grunewald @mulhearn Also added the lower-level getters for stage2 data members
  f2fc89d07579db974f0436f568837cfe0e55fe53
@mulhearn should pick it up and update his PR

**Comment by mulhearn** [comment]
Cherry-picking them now...

**Comment by mulhearn** [comment]
OK, I'm synced up to f2fc89d.

**Comment by rekovic** [comment]
Hi @Martin-Grunewald I just wanted to double check, who is going to do this ?

(BTW, what is the status of the HLT trigger summary producers wrt the new L1T objects?
@fwyzard is this something that you and @perrotta could work on ? looks like Mike and Vladimir are quite busy fixing the rest :-(  )

**Comment by rekovic** [comment]
@Martin-Grunewald @fwyzard 
I can work on this tonight if needed, is what I wanted to say.

**Comment by Martin-Grunewald** [comment]
@rekovic 
That would be great, thanks!
